doc/designs/fetching-composite-algorithms.md

```diff
@@ -0,0 +1,175 @@
+Fetching composite algorithms and using them - adding the bits still missing
+============================================================================
+
+Quick background
+----------------
+
+We currently support - at least in the public libcrypto API - explicitly
+fetching composite algorithms (such as AES-128-CBC or HMAC-SHA256), and
+using them in most cases.  In some cases (symmetric ciphers), our providers
+also provides them.
+
+However, there is one class of algorithms where the support for *using*
+explicitly fetched algorithms is lacking: asymmetric algorithms.
+
+For a longer background and explanation, see
+[Background / tl;dr](#background-tldr) at the end of this design.
+
+Public API - Add variants of `EVP_PKEY_CTX` initializers
+--------------------------------------------------------
+
+As far as this design is concerned, these API sets are affected:
+
+- SIGNATURE (DigestSign and DigestVerify)
+- ASYM_CIPHER
+- KEYEXCH
+
+The proposal is to add these functions:
+
+``` C
+EVP_DigestSignInit_ex2(EVP_PKEY_CTX **pctx,
+                       EVP_SIGNATURE *sig, EVP_PKEY *pkey,
+                       OSSL_LIB_CTX *libctx, const OSSL_PARAM params[]);
+EVP_DigestVerifyInit_ex2(EVP_PKEY_CTX **pctx,
+                         EVP_SIGNATURE *sig, EVP_PKEY *pkey,
+                         OSSL_LIB_CTX *libctx, const OSSL_PARAM params[]);
+
+int EVP_PKEY_encrypt_init_ex2(EVP_PKEY_CTX *ctx, EVP_ASYM_CIPHER,
+                              const OSSL_PARAM params[]);
+int EVP_PKEY_decrypt_init_ex2(EVP_PKEY_CTX *ctx, EVP_ASYM_CIPHER,
+                              const OSSL_PARAM params[]);
+
+int EVP_PKEY_derive_init_ex2(EVP_PKEY_CTX *ctx, EVP_KEYEXCH,
+                             const OSSL_PARAM params[]);
+```
+
+Because `EVP_SIGNATURE`, `EVP_ASYM_CIPHER` and `EVP_KEYEXCH` aren't limited
+to composite algorithms, these functions can be used just as well with
+explicit fetches of simple algorithms, say "RSA".  In that case, the caller
+will need to pass necessary auxiliary parameters through the `OSSL_PARAM` or
+a call to a corresponding `set_params` function.
+
+Requirements on the providers
+-----------------------------
+
+Because it's not immediately obvious from a composite algorithm name what
+key type it requires / supports, at least in code, allowing the use of an
+explicitly fetched implementation requires that providers cooperate by
+declaring what key type is required / supported by each algorithm.
```

Comment:
I am just wondering why we need to know this. We are passed a fetched algorithm from which we can obtain a name. We are also passed a key in the `EVP_PKEY_CTX` which has an associated keymgmt. From there we can call the keymgmt `query_operation_name` function and confirm that the name returned matches the name of the algorithm we have been passed. Isn't that sufficient?

Reply:
No.  Say that the algorithm name is "sha256WithRSAEncryption".  There's no chance that any keymgmt will return that operation name for any key.

Basically, you have to consider that there are numerous composite algorithms for any one key type.  Asking the keymgmt for the operation to be used doesn't answer how they will fit together.  Asking the composite algorithm for the key type to be used does.

Reply:
Furthermore, this would imply that for any signature, we would need to do a *search* through all available keymgmt implementations to see if any one says it supports this particular (in my example signature) algorithm.

Reply:
Hmm. Ok. Yeah that makes sense. But couldn't we do that as a fallback strategy where a provider does not supply the new query function? In that case such a provider probably does not support composite algorithms.

Reply:
> we would need to do a search through all available keymgmt implementations to see if any one says it supports this particular (in my example signature) algorithm.

Don't we just ask the keymgmt for the key we've been passed (after export to the target provider if necessary)? Why is a search necessary?

Reply:
[sigh] This section talks about *composite algorithms* specifically.

You. Will. Not. Find. A. Key. Type. Name. That. Matches. A. **Composite**.  Algorithm.  Name.
At least as far as I've seen things develop so far.

What you're talking about, I think, is cases where someone has fetched a base signature algorithm (such as "RSA") and intends to call the new initialization functions with an OSSL_PARAM array that specifies the digest to use or whatnot.  That is not what this section is about, but I hope you have the answers covering those cases in the fallback strategy section.

So I guess that I have to write a little more here to clarify that this really just is about composite algorithms.

Reply:
I did a bit or rewriting, and hope that clarifies things